keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/memory/InMemoryArtifactRepository.kt

```diff
@@ -12,44 +12,95 @@ import com.netflix.spinnaker.keel.api.PromotionStatus
 import com.netflix.spinnaker.keel.api.PromotionStatus.CURRENT
 import com.netflix.spinnaker.keel.api.PromotionStatus.DEPLOYING
 import com.netflix.spinnaker.keel.api.PromotionStatus.PREVIOUS
+import com.netflix.spinnaker.keel.persistence.ArtifactNotFoundException
 import com.netflix.spinnaker.keel.persistence.ArtifactRepository
 import com.netflix.spinnaker.keel.persistence.NoSuchArtifactException
+import java.util.UUID
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 
 class InMemoryArtifactRepository : ArtifactRepository {
-  private val artifacts = mutableMapOf<DeliveryArtifact, MutableList<ArtifactVersionAndStatus>>()
-  private val approvedVersions = mutableMapOf<Key, MutableList<String>>()
-  private val deployedVersions = mutableMapOf<Key, MutableList<String>>()
-  private val statusByEnvironment = mutableMapOf<Key, MutableMap<String, PromotionStatus>>()
+  // we want to store versions by name and type, not each artifact, so that we only store them once
+  private val versions = mutableMapOf<VersionsKey, MutableList<ArtifactVersionAndStatus>>()
```

Comment:
This is a change across both artifact repositories. I'm storing the versions of an artifact keyed off of "name" and "type". This separates the versions from the delivery artifact. I did this so that several people could care about different facets of an artifact - say, only `RELEASE` builds for a debian, and a specific versioning strategy for docker images. In the sql implementation this is the new `artifact_versions` table.

We store information about the delivery artifact separately, so that we can look up the correct filtering and sorting when we need them.